Use matroid-database package

gmou3 opened the pull request:
Remove matroid database files (`ext_data/matroids`) and use newly created `matroid-database` pypi package instead.

Links to `matroid-database`: [pypi](https://pypi.org/project/matroid-database/), [github](https://github.com/gmou3/matroid-database)

Outdated:
> Compress (via `xz -e`) the matroid database files (~75MB to <1MB). The files are now opened with `lzma.open()` and the file paths are located through `sage.ext_data` and not relative to `SAGE_SRC` or `SAGE_EXTCODE`. These changes attempt to address the issues raised by @tornaria.
> 
> Also, the addition of the matroid database files inside the folder `ext_data` needs to be noted in the file `setup.cfg.m4`. This issue was brought to my attention by @antonio-rojas.
> 
> These follow the merging of #37140. 
> 
> I also took the initiative of reordering alphabetically the list of paths inside `setup.cfg.m4` and of removing the line `ext_data/images/*` which seemed redundant.

<!-- Why is this change required? What problem does it solve? -->
<!-- If this PR resolves an open issue, please link to it here. For example "Fixes #12345". -->
<!-- If your change requires a documentation PR, please link it appropriately. -->

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.

gmou3 committed:
Update src/setup.cfg.m4 to include matroid database

gmou3 committed:
Merge branch 'develop' into all_matroids

gmou3 committed:
gzip matroid database

gmou3 committed:
Merge branch 'all_matroids' into ext_data_matroids

gmou3 committed:
Lint

tornaria commented on src/sage/matroids/database_collections.py:
```diff
@@ -195,24 +197,20 @@ def AllMatroids(n, r=None, type="all"):
             type_file = "all" if (type != "unorientable") else "unorientable"
             file = os.path.join(
-                str(SAGE_EXTCODE), "matroids", "database",
+                str(SAGE_SRC), "sage", "ext_data", "matroids", "database",
```
Using `SAGE_SRC` is not the way to avoid using `SAGE_EXTCODE`. The idea of #33037 is that one should not use neither `SAGE_LIB` nor `SAGE_SRC`, and `SAGE_EXTCODE`.

Instead, one is supposed to use `importlib.resources` to access package data (which I guess should be placed in the package that owns the data instead of the generic `src/sage/ext_data` although I guess you could still access files in there via `sage.ext_data`).

mkoeppe replied:
That's right, using `SAGE_SRC` should also be avoided completely because this variable does not have a meaningful definition in the presence of PEP 420 namespace packages.

gmou3 committed:
Compress using XZ and avoid SAGE_SRC/EXTCODE

xz with -e flag (extreme)

gmou3 committed:
Remove ext_data/matroids and use optional matroid-database pypi pkg

Newly created matroid-database pypi package: https://pypi.org/project/matroid-database/

gmou3 committed:
Reviewers' suggestions for matroid_database

Add PythonModule('matroid_database')
Use # optional - matroid_database

mkoeppe commented on src/sage/features/databases.py:
```diff
@@ -263,6 +263,7 @@ def __init__(self, name='polytopes_db'):
 
 def all_features():
     return [PythonModule('conway_polynomials'),
+            PythonModule('matroid_database'),
```
```suggestion
    return [PythonModule('conway_polynomials', spkg='conway_polynomials', type='standard'),
            PythonModule('matroid_database', spkg='matroid_database'),
```

gmou3 committed:
Update src/sage/features/databases.py

Co-authored-by: Matthias Köppe <mkoeppe@math.ucdavis.edu>

mkoeppe commented on src/setup.cfg.m4:
```diff
@@ -107,27 +107,26 @@ sage.repl.rich_output =
 
 sage =
     ext_data/*
-    ext_data/kenzo/*
```
I would suggest to back out the cosmetic change (alphabetic sorting) from this PR because it will conflict with #36951

gmou3 committed:
Revert state of src/setup.cfg.m4

gmou3 committed:
Update matroid-database to version 0.2

mkoeppe commented on build/pkgs/matroid_database/SPKG.rst:
```diff
@@ -0,0 +1,16 @@
+
+License
+-------
+
```
please add license and more detailed description -- for example by fixing the metadata of the package on PyPI and re-running `sage --package create matroid_database --pypi`

gmou3 replied:
I did so manually.

mkoeppe commented on src/sage/features/databases.py:
```diff
@@ -262,7 +262,8 @@ def __init__(self, name='polytopes_db'):
 def all_features():
-    return [PythonModule('conway_polynomials'),
+    return [PythonModule('conway_polynomials', spkg='conway_polynomials', type='standard'),
+            PythonModule('matroid_database', spkg='matroid_database'),
```
I'd suggest to make this a separate feature class `DatabaseMatroids`.

Then, just before you `import matroid_database`, use `DatabaseMatroids().require()`. This will give instructions to users how to install it.

gmou3 committed:
Add more matroid-database info

gmou3 committed:
Add class DatabaseMatroids
